File: backend_java/src/main/java/tic_tac_toe/Game.java
package tic_tac_toe;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.logging.Logger;
import tic_tac_toe.FinishException;

public class Game {

	// general constants
	final int BOARD_SIZE = 3;
	final String LINE = "line";
	final String COLUMN = "column";
	final String PLAYER = "player";
	final String TOKEN = "token";
	String O = "O";
	String X = "X";
	private static final Logger LOGGER = Logger.getLogger(Game.class.getName());

	// global variables
	public String game_id;
	public Map<String, String> players;
	private String winner;
	private ArrayList<Map> moves;

	public Game() {
		this.game_id = UUID.randomUUID().toString();
		this.players = new HashMap<String, String>();
		this.players.put(X, "");
		this.players.put(O, "");
		this.winner = null;
		this.moves = new ArrayList<Map>();
	}

	private String[][] genBoard() {
		String[][] t = { { "", "", "" }, { "", "", "" }, { "", "", "" } };
		return t;
	}

	public String[][] getBoard() {
		String[][] board = genBoard();

		for (Map<String, String> move : this.moves) {
			Integer line = Integer.parseInt(move.get(this.LINE));
			Integer colum = Integer.parseInt(move.get(this.COLUMN));
			String player = move.get(this.PLAYER);
			board[line][colum] = player;
		}
		//LOGGER.info(Arrays.deepToString(board));
		return board;
	}

	public String getWinner() {
		return this.winner;
	}

	public ArrayList<String> getPlayers() {
		ArrayList<String> ret = new ArrayList<String>();
		for (Map.Entry<String, String> entry : this.players.entrySet()) {
			if (entry.getValue().length() != 0) {
				ret.add(entry.getKey());
			}
		}
		return ret;
	}

	public String getPlayerByToken(String token) {
		for (Map.Entry<String, String> entry : this.players.entrySet()) {
			if (entry.getValue() == token) {
				return entry.getKey();
			}
		}
		return null;
	}

	private void validatePosition(int line, int column) {

		if ((line > BOARD_SIZE) || (line < 0)) {
			LOGGER.info("Line must be between 1 and {LINE_LEN}");
		}

		if ((column > BOARD_SIZE) || (column < 0)) {
			LOGGER.info("Column must be between 1 and {COLUMN_LEN}");
		}
		return;
	}

	public boolean registerMove(String l, String c, String player_id) throws GameErrorException {
		int line = Integer.parseInt(l);
		int column = Integer.parseInt(c);

		validatePosition(line, column);

		String[][] board = getBoard();

		// check if place is empty
		if (board[line][column] != "") {
			throw new GameErrorException(String.format("This position is already full with %s", board[line][column]),
					this.game_id);
		}

		// check if its players turn
		int last = this.moves.size() - 1;
		if ((this.moves.size() > 0) && (this.moves.get(last).get(PLAYER) == player_id)) {
			throw new GameErrorException("You did the last move, wait for other player", this.game_id);
		}

		LOGGER.info(String.format("%s -> board[%s][%s] = %s", this.game_id, l, c, player_id));
		return this.moves.add(Map.of(PLAYER, player_id, COLUMN, c, LINE, l));
	}

	public boolean doMove(String line, String column, String token) {

		String player_id = getPlayerByToken(token);

		if (this.getWinner() != null) {
			throw new GameErrorException("Game is already finished, can't move", this.game_id);
		}

		if (this.players.containsKey(player_id) == false) {
			throw new GameErrorException(String.format("Player %s not found", player_id), this.game_id);
		}

		// force player to uppercase
		player_id = player_id.toUpperCase();

		// this will throw exception in case of failure
		registerMove(line, column, player_id);

		// after doing some action, verify if it's done
		isFinished();

		return true;
	}

	private void isFinished() throws FinishException{
		/*
		 * In Tic tac, you can win by: 
		 *     completing a line
		 *     completing a column
		 *     completing a diagonal
		 */
		String[][] board = getBoard();

		checkLines(board);
		checkColumns(board);
		checkDiagonals(board);
		checkDraw(board);
	}

	private void checkDiagonals(String[][] board) throws FinishException {
		// check for diagonals
		Set<String> left_diagonal = new HashSet<String>();
		left_diagonal.addAll(Arrays.asList(new String[] { board[0][0], board[1][1], board[2][2] }));
		Set<String> right_diagonal = new HashSet<String>();
		right_diagonal.addAll(Arrays.asList(new String[] { board[0][2], board[1][1], board[0][2] }));

		checkSet(left_diagonal, "Left diagonal filled by player");
		checkSet(right_diagonal, "Right filled by player");
		return;
	}

	private void checkLines(String[][] board) throws FinishException {

		for (int line = 0; line < BOARD_SIZE; line++) {
			Set<String> line_set = new HashSet<String>();
			line_set.addAll(Arrays.asList(board[line]));
			checkSet(line_set, String.format("Line %d filled by player", line));
		}
		return;
	}

	private void checkColumns(String[][] board) throws FinishException {

		for (int column = 0; column < BOARD_SIZE; column++) {
			Set<String> column_set = new HashSet<String>();
			column_set.addAll(Arrays.asList(new String[] { board[0][column], board[1][column], board[2][column] }));
			checkSet(column_set, String.format("Column %d filled by player", column));
		}
		return;
	}

	private void checkSet(Set<String> thisSet, String message) throws FinishException {
		if ((thisSet.size() == 1) && (players.containsKey(thisSet.toArray()[0]))) {
			this.winner = (String) thisSet.toArray()[0];
			throw new FinishException(String.format("%s %s", message, this.winner), this.game_id);
		}
		return;
	}

	private void checkDraw(String[][] board) throws FinishException {
		for (String[] line : board) {
			for (String item : line) {
				if (item == "") {
					return;
				}
			}
		}
		throw new FinishException("Draw game", this.game_id);
	}
}

File: backend/tic_tac_toe/game.py
import uuid
from tic_tac_toe import log

logger = log.getLogger()

BOARD_SIZE = 3

LINE = "line"
COLUMN = "column"
PLAYER = "player"
TOKEN = "token"
O = 'O'
X = 'X'


class Game:
    def __init__(self):
        self._game_id = uuid.uuid4().hex
        self._players = {X: None, O: None}
        self._winner = None
        self._moves = []

    def _gen_board(self):
        board = []
        for line in range(BOARD_SIZE):
            board.append([None for column in range(BOARD_SIZE)])

        return board

    def __str__(self):
        return self._game_id

    def get_board(self):
        board = self._gen_board()

        for move in self._moves:
            line = move[LINE]
            colum = move[COLUMN]
            player = move[PLAYER]
            board[line][colum] = player

        return board

    def get_winner(self):
        return self._winner

    def get_players(self):
        return [player for player,token in self._players.items() if token is not None]

    def _get_player_by_token(self, token):
        for player,tk in self._players.items():
            if token == tk:
                return player
        raise GameException("Player not found")

    def _validate_position(self, line: int, column: int):

        if isinstance(column, int) is False or isinstance(line, int) is False:
            raise GameException(f"Position cordinates must be integer value")

        if line > BOARD_SIZE or line < 0:
            raise GameException("Line must be between 1 and {LINE_LEN}")

        if column > BOARD_SIZE or column < 0:
            raise GameException("Column must be between 1 and {COLUMN_LEN}")

    def _register_move(self, line: int, column: int, player_id: str) -> bool:
        self._validate_position(line, column)

        board = self.get_board()

        # check if place is empty
        if board[line][column] is not None:
            raise GameException(
                f"This position is already full with {board[line][column]}")

        # check if its players turn
        if len(self._moves) > 0 and self._moves[-1][PLAYER] == player_id:
            raise GameException("You did the last move, wait for other player")

        logger.debug(
            f"{self._game_id} -> board[{line}][{column}] = {player_id}")
        self._moves.append({PLAYER: player_id, COLUMN: column, LINE: line})

    def do_move(self, line: int, column: int, token: str) -> bool:
        player_id = self._get_player_by_token(token)

        if self._winner is not None:
            raise GameException("Game is already finished, can't move")

        if player_id not in self._players.keys():
            raise GameException(f"Player {player_id} not found")

        # force player to uppercase
        player_id = player_id.upper()

        # this will throw exception in case of failure
        self._register_move(line, column, player_id)

        # after doing some action, verify if it's done
        self.is_finished()

        return True

    def is_finished(self):
        """
        In Tic tac, you can win by:
            * completing a line
            * completing a colum
            * completing a diagonal
        """
        board = self.get_board()

        # check for line
        for line in board:

            line_set = list(set(line))
            if len(line_set) == 1 and line_set[0] in self._players.keys():
                self._winner = line_set[0]
                raise GameFinished(f"Line {line} completed by {self._winner}")

        # check for columns
        for column in range(BOARD_SIZE):
            colum_values = list(
                set([board[line][column] for line in range(BOARD_SIZE)]))

            if len(colum_values) == 1 and colum_values[0] in self._players.keys():
                self._winner = colum_values[0]
                raise GameFinished(
                    f"Column {column} completed by {self._winner}")

        # check for diagonals
        diagonal_left = list(
            set([board[index][index] for index in range(BOARD_SIZE)]))
        if len(diagonal_left) == 1 and diagonal_left[0] in self._players.keys():
            self._winner = diagonal_left[0]
            raise GameFinished(f"Left diagonal completed by {self._winner}")

        diagonal_right = list(
            set([
                board[index][(BOARD_SIZE - 1) - index]
                for index in range(BOARD_SIZE)
            ]))
        if len(diagonal_right) == 1 and diagonal_right[0] in self._players.keys():
            self._winner = diagonal_right[0]
            raise GameFinished(f"Right diagonal completed by {self._winner}")


class APIException(Exception):
    status_code = 400

    def __init__(self, message, status_code=None, payload=None):
        Exception.__init__(self)
        self.message = message
        if status_code is not None:
            self.status_code = status_code
        self.payload = payload
        logger.debug(message)

    def to_dict(self):
        rv = dict(self.payload or ())
        rv['message'] = self.message
        return rv


class GameException(APIException):
    status_code = 400


class GameFinished(APIException):
    status_code = 218
